src/object/C4GameObjects.cpp

```diff
@@ -46,355 +46,401 @@ void C4GameObjects::Default()
 		{
 			uint32_t Marker = GetNextMarker();
-			C4LSector *pSct;
-			for (C4ObjectList *pLst = obj1->Area.FirstObjects(&pSct); pLst; pLst = obj1->Area.NextObjects(pLst, &pSct))
-				for (C4Object* obj2 : *pLst)
-					if ((obj2 != obj1) && obj2->Status && !obj2->Contained && (obj2->OCF & tocf) &&
-					    Inside<int32_t>(obj2->GetX() - (obj1->GetX() + obj1->Shape.x), 0, obj1->Shape.Wdt - 1) &&
-					    Inside<int32_t>(obj2->GetY() - (obj1->GetY() + obj1->Shape.y), 0, obj1->Shape.Hgt - 1) &&
-					    obj1->Layer == obj2->Layer)
+			C4LSector *sector;
+			for (C4ObjectList *in_goal_area = goal->Area.FirstObjects(&sector); in_goal_area; in_goal_area = goal->Area.NextObjects(in_goal_area, &sector))
+			{
+				for (C4Object* ball : *in_goal_area)
+				{
+					if ((ball != goal) && ball->Status && !ball->Contained && (ball->OCF & ball_required_ocf) &&
+					    Inside<int32_t>(ball->GetX() - (goal->GetX() + goal->Shape.x), 0, goal->Shape.Wdt - 1) &&
+					    Inside<int32_t>(ball->GetY() - (goal->GetY() + goal->Shape.y), 0, goal->Shape.Hgt - 1) &&
+					    goal->Layer == ball->Layer)
 					{
-						// handle collision only once
-						if (obj2->Marker == Marker) continue;
-						obj2->Marker = Marker;
+						// Handle collision only once
+						if (ball->Marker == Marker)
+						{
+							continue;
+						}
+						ball->Marker = Marker;
 						// Only hit if target is alive and projectile is an object
-						if ((obj1->OCF & OCF_Alive) && (obj2->Category & C4D_Object))
+						if ((goal->OCF & OCF_Alive) && (ball->Category & C4D_Object))
 						{
-							C4Real dXDir = obj2->xdir - obj1->xdir, dYDir = obj2->ydir - obj1->ydir;
-							C4Real speed = dXDir * dXDir + dYDir * dYDir;
+							C4Real relative_xdir = ball->xdir - goal->xdir;
+							C4Real relative_ydir = ball->ydir - goal->ydir;
+							C4Real speed = relative_xdir * relative_xdir + relative_ydir * relative_ydir;
 							// Only hit if obj2's speed and relative speeds are larger than HitSpeed2
-							if ((obj2->OCF & OCF_HitSpeed2) && speed > HitSpeed2 &&
-							   !obj1->Call(PSF_QueryCatchBlow, &C4AulParSet(obj2)))
+							if ((ball->OCF & OCF_HitSpeed2) && speed > HitSpeed2 &&
+							   !goal->Call(PSF_QueryCatchBlow, &C4AulParSet(ball)))
 							{
-								int32_t iHitEnergy = fixtoi(speed * obj2->Mass / 5);
+								int32_t hit_energy = fixtoi(speed * ball->Mass / 5);
 								// Hit energy reduced to 1/3rd, but do not drop to zero because of this division
-								iHitEnergy = std::max<int32_t>(iHitEnergy/3, !!iHitEnergy);
-								obj1->DoEnergy(-iHitEnergy / 5, false, C4FxCall_EngObjHit, obj2->Controller);
-								int tmass = std::max<int32_t>(obj1->Mass, 50);
-								C4PropList* pActionDef = obj1->GetAction();
+								hit_energy = std::max<int32_t>(hit_energy/3, !!hit_energy);
+								goal->DoEnergy(-hit_energy / 5, false, C4FxCall_EngObjHit, ball->Controller);
+								int tmass = std::max<int32_t>(goal->Mass, 50);
```

Comment:
I think `tmass` should be `goal_mass` with your naming scheme.

Reply:
Done.